### PR TITLE
[Relay] Strict mode in pattern matching

### PR DESCRIPTION
**Pull request by MarisaKirisame**
the strict mode require that all pattern is exhausted. by default it is on, so one had to explicitly turn it off for hd, tl, foldl1, etc. This stop all the warning generated by the exhaust matcher, and enforce better coding style as partial function will be more explicit.

### CODE REVIEW

**Comment by slyubomirsky** [comment]
Very good addition, thanks for doing it

**Comment by slyubomirsky** [comment]
Should the match completeness and the fatal stuff be the same PR, though? There are more issues where fatal is involved:
- Ensuring interpreters handle it and testing it
- Ensuring passes that may need to care about it do handle it

**Comment by MarisaKirisame** [comment]
We'll do fatal next time

**Comment by slyubomirsky** [comment]
Thanks for removing fatal, I think this is a good change

File: include/tvm/relay/adt.h [hunk 1]
```diff
@@ -241,14 +241,18 @@ class MatchNode : public ExprNode {
   /*! \brief The match node clauses. */
   tvm::Array<Clause> clauses;
 
+  /*! \brief Is the match complete (cover all cases)? */
```
Comment (slyubomirsky) [inline include/tvm/relay/adt.h]: Small nitpick: I think the comment here and in `adt.py` should be a little clearer about what the variable means: It means whether the match _should_ be complete, not whether it _is_ complete (which the typechecker checks).

```suggestion
  /*! \brief Should this match be complete (cover all cases)? 
   *  If yes, the type checker will warn about any missing cases.
   */
```

File: python/tvm/relay/adt.py [hunk 2]
```diff
@@ -186,18 +186,23 @@ def __init__(self, lhs, rhs):
...
         Parameters
         ----------
         data: tvm.relay.Expr
             The value being deconstructed and matched.
+
         clauses: List[tvm.relay.Clause]
             The pattern match clauses.
+
+        complete: Optional[Bool]
+            Is the pattern complete(cover all cases)?
```
Comment (slyubomirsky) [inline python/tvm/relay/adt.py]:
```suggestion
            Should the match be complete (cover all cases)?
            If yes, the type checker will warn about any missing cases 
```
Analogous to my other comment

Comment (MarisaKirisame) [reply python/tvm/relay/adt.py]: It is now an error. Can you change the warn to the error? I sort of dont have a enviroment right now... Thx.

Comment (slyubomirsky) [reply python/tvm/relay/adt.py]: I updated my suggestions

**Pull request merged**